### PR TITLE
box props for radiobuttongroup added in theme

### LINKED ISSUE #4184: Would like ability for RadioButtonGroup styling to be set in theme
### Expected Behavior

I'd like to be able to specify styling for RadioButtonGroup in a theme, specifically the options `gap`.

### Actual Behavior

Theme not currently referenced.
RadioButtonGroup styling can not be set in theme (e.g. desired gap size)
  - Same applies to CheckBoxGroup

Use case. HPE Design System designs desire a gap of none between RadioButtonGroup and CheckBoxGroup options. Implementers can achieve this by setting the `gap` property for each instance, however for the Design System, this specification would make sense to live in the theme.

### Steps to Reproduce

1. 2. 3.

### Your Environment

- Grommet version:
- Browser Name and version:
- Operating System and version (desktop or mobile):

### PR DESCRIPTION
**Pull request by IanKBovard**
#### What does this PR do?
closes #4184 
Adds a gap to RadioButtonGroup in theme

#### Where should the reviewer start?
RadioButtonGroup.js
base.js
base.d.ts
RadioButtonGroup Stories - CustomTheme.js
RadioButtonGroup doc.js

#### What testing has been done on this PR?
Storybook, tested with and without custom theme.
Used none, xxsmall, small, medium, large, xlarge, and px values to test gap within custom theme.

#### How should this be manually tested?
same as above

#### Any background context you want to provide?
RadioButtonGroup did not have the ability to set gap within theme.

#### What are the relevant issues?
#4341 

#### Do the grommet docs need to be updated?
Yes, they are in this pr

#### Is this change backwards compatible or is it a breaking change?
backwards compatible

### CODE REVIEW

File: src/js/components/RadioButtonGroup/RadioButtonGroup.js [hunk 1]
```diff
@@ -27,6 +28,10 @@ const RadioButtonGroup = forwardRef(
     ref,
   ) => {
     const formContext = useContext(FormContext);
+    const theme = useContext(ThemeContext) || defaultProps.theme;
+    const gap = theme.global.radioButtonGroup
+      ? theme.global.radioButtonGroup.gap
+      : 'small';
```
Comment (ShimiSun) [inline src/js/components/RadioButtonGroup/RadioButtonGroup.js:34]: Would setting a default value here will be backward compatible?

File: src/js/components/RadioButtonGroup/doc.js [hunk 2]
```diff
@@ -61,3 +61,22 @@ export const doc = RadioButtonGroup => {
 
   return DocumentedRadioButtonGroup;
 };
+
+export const themeDoc = {
+  'radioButtonGroup.gap': {
+    description: `The amount of spacing between child elements. This
+    should not be used in conjunction with 'wrap' as the gap elements
+    will not wrap gracefully.`,
```
Comment (ShimiSun) [inline src/js/components/RadioButtonGroup/doc.js:69]: Not sure this level of description is relevant for a theme prop, thoughts?

File: src/js/themes/base.d.ts [hunk 3]
```diff
@@ -874,7 +887,10 @@ export interface ThemeType {
     size?: string;
     font?: {
       weight?: number | string;
-    }
+    };
+  };
+  radioButtonGroup?: {
+    gap?: string;
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts:893]: I think it should be GapType instead of string. import it from utils.

File: src/js/components/RadioButtonGroup/RadioButtonGroup.js (hunk 1 above)
Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/RadioButtonGroup.js:34]: in the original code radiobuttongroup had a default value of small.
```
const RadioButtonGroup = forwardRef(
  (
    {
      children,
      disabled,
      gap = 'small',
      name,
      onChange,
      options: optionsProp,
      value: valueProp,
      ...rest
    },
    ref,
  ) => {
```

File: src/js/components/RadioButtonGroup/doc.js (hunk 2 above)
Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/doc.js:69]: its just copied from box, i can remove that section

File: src/js/themes/base.d.ts (hunk 3 above)
Comment (IanKBovard) [reply src/js/themes/base.d.ts]: done

File: src/js/components/RadioButtonGroup/doc.js (hunk 2 above)
Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/doc.js]: removed that section

File: src/js/components/RadioButtonGroup/RadioButtonGroup.js (hunk 1 above)
Comment (ShimiSun) [reply src/js/components/RadioButtonGroup/RadioButtonGroup.js:34]: We should keep the behavior of the original code and add the theme option for the `gap` without setting a default value on it. That will be more backward compatible in case a user is overriding the theme.

**Comment by britt6612** [comment]
This code looks fine 
However I think the original issue was for the vertical gap between the icon on radio button and the label.
this can be done with gap=“none”, but would require developers to always set that prop. So the intent was for the gap to be specified in the theme. This was so in the design-system we can control this in the theme

```
<RadioButtonGroup
          id="power-stzw"
          name="power-regulation"
          options={powerRegulationOptions}
          value={powerRegulation}
          onChange={event => setPowerRegulation(event.target.value)}
          gap="none"
        />
```

File: src/js/components/RadioButtonGroup/RadioButtonGroup.js (hunk 1 above)
Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/RadioButtonGroup.js]: changes made

File: src/js/themes/base.d.ts [hunk 4]
```diff
@@ -81,10 +81,12 @@ type Colors = typeof colors & {
 
 interface ButtonKindType {
   background?: BackgroundType;
-  border?: {
-    color?: ColorType;
-    width?: string;
-  } | boolean;
+  border?:
```
Comment (britt6612) [inline src/js/themes/base.d.ts:84]: Maybe change linter back to how it is

Comment (britt6612) [reply src/js/themes/base.d.ts:84]: This add extra pipe before

Comment (IanKBovard) [reply src/js/themes/base.d.ts:84]: when i commit it auto lints it for me. is there a config im missing?

Comment (britt6612) [reply src/js/themes/base.d.ts:84]: hmm maybe that got changed ? @ShimiSun is this the right linting ?

Comment (ShimiSun) [reply src/js/themes/base.d.ts:84]: We recently added linter & prettier for typescript files. So I think it might be a side effect, we'll need to take a look whether that is the default behavior that is a best practice.

Comment (IanKBovard) [reply src/js/themes/base.d.ts:84]: do i need a new config file, or to adjust the current config to account for this?

Comment (ShimiSun) [reply src/js/themes/base.d.ts:84]: It will probably be a separate task from this PR, leave it be for now if its complaint with lint.

File: src/js/components/RadioButtonGroup/RadioButtonGroup.js [hunk 5]
```diff
@@ -27,6 +29,10 @@ const RadioButtonGroup = forwardRef(
     ref,
   ) => {
     const formContext = useContext(FormContext);
+    const theme = useContext(ThemeContext) || defaultProps.theme;
+    const radioButtonGap = theme.global.radioButtonGroup
+      ? theme.global.radioButtonGroup.gap
+      : gap;
```
Comment (ShimiSun) [inline src/js/components/RadioButtonGroup/RadioButtonGroup.js:35]: I think that instead of fetching only the `gap` style from the theme, the behavior here should be more generic and should fetch all the Box props from the theme.
So essentially this const will be removed, and see my changes on line 110.

File: src/js/components/RadioButtonGroup/RadioButtonGroup.js [hunk 6]
```diff
@@ -101,7 +107,7 @@ const RadioButtonGroup = forwardRef(
         onLeft={focus ? onPrevious : undefined}
         onRight={focus ? onNext : undefined}
       >
-        <Box ref={ref} gap={gap} {...rest}>
+        <Box ref={ref} gap={radioButtonGap} {...rest}>
```
Comment (ShimiSun) [inline src/js/components/RadioButtonGroup/RadioButtonGroup.js:110]:
```suggestion
        <Box ref={ref} {...theme.radioButtonGroup} {...rest}>
```
This change will enable all Box props and not just `gap`.

Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/RadioButtonGroup.js]: changes made

File: src/js/components/RadioButtonGroup/doc.js [hunk 7]
```diff
@@ -61,3 +61,20 @@ export const doc = RadioButtonGroup => {
...
+    defaultValue: `{
+      none,
+      xxsmall,
+      xsmall,
+      small,
+      medium,
+      large,
+      xlarge,
+      any css size,
+    }`,
```
Comment (ShimiSun) [inline src/js/components/RadioButtonGroup/doc.js:78]: I don't think the default value is an object, it should be `undefined`.

Comment (ShimiSun) [reply src/js/components/RadioButtonGroup/doc.js:78]: And regardless, the docs need to accommodate your recent changes where all the Box props are included and not a single doc entry just for gap prop.

File: src/js/themes/base.d.ts [hunk 8]
```diff
@@ -874,7 +887,10 @@ export interface ThemeType {
     size?: string;
     font?: {
       weight?: number | string;
-    }
+    };
+  };
+  radioButtonGroup?: {
+    gap?: GapType;
```
Comment (ShimiSun) [inline src/js/themes/base.d.ts:893]: This should accommodate your recent changes.

File: src/js/themes/base.js [hunk 9]
```diff
@@ -909,6 +909,9 @@ export const generate = (baseSpacing = 24, scale = 6) => {
         // weight: undefined,
       },
     },
+    radioButtonGroup: {
+      // gap: undefined,
```
Comment (ShimiSun) [inline src/js/themes/base.js:913]: should include your recent changes.

File: src/js/themes/base.d.ts (hunk 8 above)
Comment (IanKBovard) [reply src/js/themes/base.d.ts:893]: gotcha, just to clarify, something like this:

```
radioButtonGroup?: {
  a11yTitle: string;
  align: string;
  alignContent: string;
  ...
  ...
}
```

File: src/js/components/RadioButtonGroup/doc.js (hunk 7 above)
Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/doc.js:78]: something like this:
```
export const themeDoc = {
  'radioButtonGroup.a11yTitle': {
    description: `a user friendly label for screen readers`,
    type: 'string',
    defaultValue: `undefined`,
  },
'radioButtonGroup.align': {
    description: `How to align the contents along the cross axis.`,
    type: 'string',
    defaultValue: `undefined`,
  }
  ...
  ...
};
```

am i correct to assume that, if there is no bolded value in the docs it defaults to undefined?

Comment (ShimiSun) [reply src/js/components/RadioButtonGroup/doc.js:78]: Similar to https://v2.grommet.io/select#select.options.container

File: src/js/themes/base.d.ts (hunk 8 above)
Comment (ShimiSun) [reply src/js/themes/base.d.ts:893]: Similar to: 
```
  card?: {
    container?: BoxProps;
  },
```

**Comment by IanKBovard** [comment]
changed made

File: src/js/components/RadioButtonGroup/doc.js [hunk 10]
```diff
@@ -61,3 +61,11 @@ export const doc = RadioButtonGroup => {
 
   return DocumentedRadioButtonGroup;
 };
+
+export const themeDoc = {
+  'radioButtonGroup.container': {
+    description: 'Any valid Box props for the RadioButtonGroup container.',
+    type: 'object',
+    defaultValue: "{ align: 'start', pad: 'small' , ...}",
```
Comment (ShimiSun) [inline src/js/components/RadioButtonGroup/doc.js:69]: Are you sure this is the default value? I'm not sure it is.

File: src/js/themes/base.js [hunk 11]
```diff
@@ -918,6 +918,11 @@ export const generate = (baseSpacing = 24, scale = 6) => {
         // weight: undefined,
       },
     },
+    radioButtonGroup: {
+      container: {
+        // gap: undefined
```
Comment (ShimiSun) [inline src/js/themes/base.js:923]: Should be commented out as well.

File: src/js/utils/index.d.ts [hunk 12]
```diff
@@ -50,7 +50,11 @@ export { isObject, deepFreeze, deepMerge, removeUndefined };
  * type SomeComponentProps = PropsOf<typeof SomeComponent>;
  * ```
  */
-export type PropsOf<TComponent> = TComponent extends React.ComponentType<infer P> ? P : never;
+export type PropsOf<TComponent> = TComponent extends React.ComponentType<
+  infer P
+>
+  ? P
+  : never;
```
Comment (ShimiSun) [inline src/js/utils/index.d.ts:57]: Please revert this change unless you changed the code's meaning.

File: src/js/components/RadioButtonGroup/doc.js (hunk 10 above)
Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/doc.js:69]: used the same description for select 
https://v2.grommet.io/select#select.options.container

```
  'select.options.container': {
    description: 'Any valid Box prop for the options container.',
    type: 'object',
    defaultValue: "{ align: 'start', pad: 'small' }",
  },
```

my assumption is that the description on the select page is not default values either, just showing values that you can pass into select.options.container.

File: src/js/utils/index.d.ts (hunk 12 above)
Comment (IanKBovard) [reply src/js/utils/index.d.ts:57]: I believe this is because of the linter.

File: src/js/themes/base.js (hunk 11 above)
Comment (IanKBovard) [reply src/js/themes/base.js]: done

File: src/js/components/RadioButtonGroup/doc.js (hunk 10 above)
Comment (ShimiSun) [reply src/js/components/RadioButtonGroup/doc.js:69]: Please put the correct default values for the RBG, and take a look at how the Select entry could be fixed as well.

File: src/js/utils/index.d.ts (hunk 12 above)
Comment (ShimiSun) [reply src/js/utils/index.d.ts:57]: Please revert it, my linter doesn't apply it, might be something with your IDEA.

File: src/js/themes/base.js [hunk 13]
```diff
@@ -918,6 +918,11 @@ export const generate = (baseSpacing = 24, scale = 6) => {
         // weight: undefined,
       },
     },
+    radioButtonGroup: {
+      // container: {
+      // gap: undefined
+      // },
```
Comment (ShimiSun) [inline src/js/themes/base.js:924]:
```suggestion
      // container: {}, // any box props
```

File: src/js/components/RadioButtonGroup/doc.js (hunk 10 above)
Comment (IanKBovard) [reply src/js/components/RadioButtonGroup/doc.js]: changed default to an empty object

File: src/js/themes/base.js (hunk 13 above)
Comment (IanKBovard) [reply src/js/themes/base.js]: made changes

File: src/js/components/RadioButtonGroup/doc.js [hunk 14]
```diff
@@ -61,3 +61,11 @@ export const doc = RadioButtonGroup => {
 
   return DocumentedRadioButtonGroup;
 };
+
+export const themeDoc = {
+  'radioButtonGroup.container': {
+    description: 'Any valid Box props for the RadioButtonGroup container.',
+    type: 'object',
+    defaultValue: '{}',
```
Comment (ShimiSun) [inline src/js/components/RadioButtonGroup/doc.js:69]:
```suggestion
    defaultValue: 'undefined',
```

File: src/js/utils/index.d.ts [hunk 15]
```diff
@@ -1,249 +0,0 @@
-// colors.js
```
Comment (ShimiSun) [inline src/js/utils/index.d.ts:1]: Why did you delete this file?

**Comment by IanKBovard** [comment]
created a new pr that fixes the linting issues on this one https://github.com/grommet/grommet/pull/4383
